relayer/processor/path_processor_internal.go

```diff
@@ -685,8 +686,30 @@ func (pp *PathProcessor) assembleAndSendMessages(
 	src, dst *pathEndRuntime,
 	messages pathEndMessages,
 ) error {
+	var needsClientUpdate bool
 	if len(messages.packetMessages) == 0 && len(messages.connectionMessages) == 0 && len(messages.channelMessages) == 0 {
-		return nil
+		var consensusHeightTime time.Time
+		if dst.clientState.ConsensusTime.IsZero() {
+			h, err := src.chainProvider.QueryIBCHeader(ctx, int64(dst.clientState.ConsensusHeight.RevisionHeight))
+			if err != nil {
+				return fmt.Errorf("failed to get header height: %w", err)
+			}
+			consensusHeightTime = time.Unix(0, int64(h.ConsensusState().GetTimestamp()))
+		} else {
+			consensusHeightTime = dst.clientState.ConsensusTime
+		}
+		if (dst.clientState.TrustingPeriod.Milliseconds() - time.Since(consensusHeightTime).Milliseconds()) < pp.clientUpdateThresholdTime.Milliseconds() {
+			needsClientUpdate = true
+			pp.log.Info("client close to expiration",
+				zap.String("chainID:", dst.info.ChainID),
+				zap.String("clientID:", dst.info.ClientID),
+				zap.Int64("trusting_period", dst.clientState.TrustingPeriod.Milliseconds()),
+				zap.Int64("time_since_client_update", time.Since(consensusHeightTime).Milliseconds()),
+				zap.Int64("client_threshold_time", pp.clientUpdateThresholdTime.Milliseconds()),
+			)
+		} else {
+			return nil
+		}
```

Comment:
So it looks like we are checking if the cached client state has a correlating consensus time cached and if not we query for the block header at the same height as the cached client state to have access to the consensus time for when this cached client state was updated last. 

Afterwards we calculate the time since the last client update by taking `time.Since(consensusHeightTime)`, which is saying `time.Now() - the last time the client was updated`, and subtract that value from the trusting period to determine if the client is going to expire within our update threshold time. And if this value is within our update threshold time we do some logging and signal that a `MsgUpdateClient` needs to be sent out.

Because this is a critical code path that users will be putting a lot of trust in, I just want to be sure I have a good mental model of how this works. It seems beneficial to document how the relayer is using the client update threshold time to determine when clients need updating. Without reading the code I would first imagine that the relayer is just sending out the `MsgUpdateClient` once every update threshold period regardless of when the last update occurred, but in reality we are determining if **ANY** relayer has updated the client and we are ensuring that the client is updated at most once within this update threshold period in relation to the trusting period.

(sorry for the wall of text)

Reply:
> Without reading the code I would first imagine that the relayer is just sending out the MsgUpdateClient once every update threshold period regardless of when the last update occurred

This was my intuition as well. But we're doing it this different way as you pointed out. 

Dan authored this part of the code and it was my understanding he paired with Andrew on it. Therefore, I didn't refactor or question it. I only added tests that hopefully cover this feature. 

You ask a great question. Given it's critical path, do you think we wait until Dan is back before merging? (I'm fine either way.)

Reply:
I think the way Dan implemented this probably makes more sense since you occur less fees by only updating the client when it's within some threshold of expiring vs. spamming the update msgs on a timer. We probably just want to document this somewhere though for users as well as any maintainers that may come in further down the line!

Reply:
I agree. Would you be ok with a quick followup PR for Dan to add documentation? He will be able to write that better than I.

Reply:
Sounds good to me! PR approved ✅

Reply:
When @boojamya and I initially paired on this, I was unaware of the update threshold period flag, so we coded it to update the clients if 2/3 of the trusting period had passed since the last update, without any configuration ability.

The flag duration in combination with the time until client expiration seems a bit confusing, so my initial thought was that we should pick one or the other. But, in today's standup, @jtieri brought up a good point that it is desirable to have both approaches since some parties may want the clients updated on a steady frequency while others may just want to ensure the clients don't expire but avoid spending unnecessary fees and only do it if necessary.

With that taken into account, I think this would be easier to understand as:
- Update client if time until client expiration is <= 1/3 of trusting period.
- Update client if time since last update is greater than update threshold period flag duration (this could default to 0 duration in which case this check would be disabled)

This would enable automatic updating within trusting period without any configuration, but would also enable more frequent client updates if desired. I think this make the flag easier to understand as an operator because, for example, an update threshold period of `2h` would update the clients every 2 hours if there is no other IBC traffic and 2 hours is less than 2/3 of the client trusting period. If the operator accidentally configured the update threshold period to be greater than the trusting period, it has the failsafe that wouldn't let the clients expire since it would still update them if not updated within 2/3 of the trusting period.